rust/otap-dataflow/crates/controller/src/thread_task.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Utilities to run a non-Send async task on a dedicated OS thread with a
+//! single-threaded Tokio runtime and LocalSet, plus a shutdown signal.
+
+use std::future::Future;
+use std::thread;
+use tokio::{runtime::Builder as RtBuilder, task::LocalSet};
+use tokio_util::sync::CancellationToken;
+
+/// Handle to a task running on a dedicated thread.
+///
+/// - `shutdown()` requests cancellation via the token (idempotent, best-effort).
+/// - `shutdown_and_join()` requests shutdown and then waits for completion, returning controller::Error on failure.
+/// - `join_raw()` waits for the thread to finish and returns the raw nested result.
+pub struct ThreadLocalTaskHandle<T, E> {
+    cancel_token: CancellationToken,
+    join_handle: Option<thread::JoinHandle<Result<T, E>>>,
+    name: String,
+}
+
+impl<T, E> ThreadLocalTaskHandle<T, E> {
+    /// Request a graceful shutdown by cancelling the token.
+    pub fn shutdown(&mut self) {
```

Comment:
It would be nice to implement `Drop` trait for `ThreadLocalTaskHandle` so that cancellation and thread join are not dependent shutdown being called.

Reply:
Done